### ISSUE TITLE
[JENKINS-3154] dangerous bug with buildWithParameters option.

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/jenkins#14173]
<p>I have created a single configuration project with "This build is parameterized"<br/>
option. <br/>
o VARIABLE NAME : FOO<br/>
o goto the shell script and echo $FOO;sleep 120</p>

<p>Now saved the configuration and started the build using following URL...</p>

<p><a href="http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST1" class="external-link" target="_blank" rel="nofollow noopener">http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST1</a></p>

<p>Now build #1 started.</p>

<p>While the build #1 is going on, I have started few more builds.<br/>
<a href="http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST2" class="external-link" target="_blank" rel="nofollow noopener">http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST2</a><br/>
<a href="http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST3" class="external-link" target="_blank" rel="nofollow noopener">http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST3</a><br/>
<a href="http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST4" class="external-link" target="_blank" rel="nofollow noopener">http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST4</a><br/>
<a href="http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST5" class="external-link" target="_blank" rel="nofollow noopener">http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST5</a><br/>
<a href="http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST6" class="external-link" target="_blank" rel="nofollow noopener">http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST6</a></p>

<p>When the build #1 is completed, build #2 & build #3 are started together. </p>

<p>When the build #2 & #3 completed, build #4 & #5 are getting started together.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vijayj">vijayj</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3154" target="_blank">dangerous bug with buildWithParameters option.</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=huybrechts">huybrechts</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Blocker
<li><b>component(s)</b>: core
<li><b>resolution</b>: Duplicate
<li><b>resolved</b>: 2009-07-09T14:17:48+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 0
<li><b>imported</b>: 2025-11-24
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>I have created a single configuration project with "This build is parameterized"
option. 
o VARIABLE NAME : FOO
o goto the shell script and echo $FOO;sleep 120</p>

<p>Now saved the configuration and started the build using following URL...</p>

<p><a href="http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST1" class="external-link" target="_blank" rel="nofollow noopener">http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST1</a></p>

<p>Now build #1 started.</p>

<p>While the build #1 is going on, I have started few more builds.
<a href="http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST2" class="external-link" target="_blank" rel="nofollow noopener">http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST2</a>
<a href="http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST3" class="external-link" target="_blank" rel="nofollow noopener">http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST3</a>
<a href="http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST4" class="external-link" target="_blank" rel="nofollow noopener">http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST4</a>
<a href="http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST5" class="external-link" target="_blank" rel="nofollow noopener">http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST5</a>
<a href="http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST6" class="external-link" target="_blank" rel="nofollow noopener">http://localhost:8080/job/buildWithParameters/buildWithParameters?FOO=TEST6</a></p>

<p>When the build #1 is completed, build #2 &amp; build #3 are started together. </p>

<p>When the build #2 &amp; #3 completed, build #4 &amp; #5 are getting started together.</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Platform: All, OS: All</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>vijayj:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3154?focusedId=129183&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-129183">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Here is the scenario for multi configuration project:</p>

<p>o Currently the build is running for multi configuration project build #150
o I have submitted new build request through buildWithParameters and that
created build #151 in the Queue
(buildWithParameters?CHANGEID=95111&amp;token=XXXXXXXXXXXXX).
o One more requested submitted through buildWithParameters and that created
build #152 in the Queue (buildWithParameters?CHANGEID=95112&amp;token=XXXXXXXXXXXXX).
o Now the build #150 completed:</p>
<ul class="alternate" type="square">
	<li>build #151 &amp; build #152 are removed from Queue and started the builds for
#151 &amp; 152 together.</li>
</ul>


<p>o #151: Within the few mins: Build #151 only executes master thread and the
marks the build as aborted but triggered all the slave process and they are
waiting in the Queue.
        &lt;log&gt;
                Triggering linux
                Triggering windows
                Triggering solaris
                linux appears to be cancelled
                windows appears to be cancelled
                solaris appears to be cancelled
                finished: ABORTED
        &lt;/log&gt;</p>

<p>O#152: Build #152 completes the master thread and starts the slave process.</p>
<ul class="alternate" type="square">
	<li>Now all the slave builds runs twice.</li>
	<li>First runs with CHANGEID=95111</li>
	<li>Second it runs with CHANGEID=95112</li>
	<li>Now the #152 has the artifacts for both #151 &amp; 152.</li>
</ul>

</pre>
</details>

</ul>
</details>

<p>Here is the scenario for multi configuration project:</p>

<p>o Currently the build is running for multi configuration project build #150<br/>
o I have submitted new build request through buildWithParameters and that<br/>
created build #151 in the Queue<br/>
(buildWithParameters?CHANGEID=95111&token=XXXXXXXXXXXXX).<br/>
o One more requested submitted through buildWithParameters and that created<br/>
build #152 in the Queue (buildWithParameters?CHANGEID=95112&token=XXXXXXXXXXXXX).<br/>
o Now the build #150 completed:</p>
<ul class="alternate" type="square">
	<li>build #151 & build #152 are removed from Queue and started the builds for<br/>
#151 & 152 together.</li>
</ul>


<p>o #151: Within the few mins: Build #151 only executes master thread and the<br/>
marks the build as aborted but triggered all the slave process and they are<br/>
waiting in the Queue.<br/>
<log><br/>
Triggering linux<br/>
Triggering windows<br/>
Triggering solaris<br/>
linux appears to be cancelled<br/>
windows appears to be cancelled<br/>
solaris appears to be cancelled<br/>
finished: ABORTED<br/>
</log></p>

<p>O#152: Build #152 completes the master thread and starts the slave process.</p>
<ul class="alternate" type="square">
	<li>Now all the slave builds runs twice.</li>
	<li>First runs with CHANGEID=95111</li>
	<li>Second it runs with CHANGEID=95112</li>
	<li>Now the #152 has the artifacts for both #151 & 152.</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mindless.png" title="mindless's avatar" /> mindless:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3154?focusedId=129184&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-129184">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>watch the console output for 2 instances that have both been assigned an
executor.. do both actually echo $FOO before either one ends, or does one block
the other until in completes and then the next one echoes $FOO?</p>
</pre>
</details>

</ul>
</details>

<p>watch the console output for 2 instances that have both been assigned an<br/>
executor.. do both actually echo $FOO before either one ends, or does one block<br/>
the other until in completes and then the next one echoes $FOO?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>vijayj:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3154?focusedId=129185&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-129185">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>For single configuration:
=========================</p>

<p>&gt; does one block the other until in completes and then the next one echoes $FOO?</p>

<p>2nd one gets blocked until the first one echoes $FOO &amp; completes. </p>

<p>For Multi configuration:
========================</p>

<p>The problem is very major with multi configuration, because both the builds
completes the master thread (not sure about order of the console out for master
thread) then triggers the slave process almost at the same time but some how the
first build is getting marked as aborted. Second build go onto execute slave
threads twice for the same build (2nd)...</p>
</pre>
</details>

</ul>
</details>

<p>For single configuration:<br/>
=========================</p>

<p>> does one block the other until in completes and then the next one echoes $FOO?</p>

<p>2nd one gets blocked until the first one echoes $FOO & completes. </p>

<p>For Multi configuration:<br/>
========================</p>

<p>The problem is very major with multi configuration, because both the builds<br/>
completes the master thread (not sure about order of the console out for master<br/>
thread) then triggers the slave process almost at the same time but some how the<br/>
first build is getting marked as aborted. Second build go onto execute slave<br/>
threads twice for the same build (2nd)...</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3154?focusedId=129186&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-129186">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I thought Tom fixed this in a very recent version of Hudson.
Which version are you using?</p>
</pre>
</details>

</ul>
</details>

<p>I thought Tom fixed this in a very recent version of Hudson.<br/>
Which version are you using?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>huybrechts:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3154?focusedId=129187&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-129187">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>When two builds of the same project are in the queue, then they will be 
scheduled at the same time. This occupies a node unnecessarily, but otherwise 
locking on the workspace is still working fine, so they don't actually run in 
parallel.</p>

<p>Maybe there is some strange interaction with matrix builds, but I still don't 
think that two builds can actually run in parallel. I'd have to try it out, or 
better yet, write some test cases.</p>
</pre>
</details>

</ul>
</details>

<p>When two builds of the same project are in the queue, then they will be <br/>
scheduled at the same time. This occupies a node unnecessarily, but otherwise <br/>
locking on the workspace is still working fine, so they don't actually run in <br/>
parallel.</p>

<p>Maybe there is some strange interaction with matrix builds, but I still don't <br/>
think that two builds can actually run in parallel. I'd have to try it out, or <br/>
better yet, write some test cases.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>vijayj:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3154?focusedId=129188&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-129188">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>&gt; Which version are you using?</p>

<p>I was using 1.285.</p>
</pre>
</details>

</ul>
</details>

<p>> Which version are you using?</p>

<p>I was using 1.285.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>vijayj:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3154?focusedId=129189&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-129189">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>&gt; Maybe there is some strange interaction with matrix builds,</p>

<p>We definitely have big issue with matrix project. If you could check this
scenario then it would be really great.</p>
</pre>
</details>

</ul>
</details>

<p>> Maybe there is some strange interaction with matrix builds,</p>

<p>We definitely have big issue with matrix project. If you could check this<br/>
scenario then it would be really great.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3154?focusedId=129190&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-129190">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I believe it's the same issue.</p>

<ul>
	<li>
	<ul>
		<li>
		<ul>
			<li>This issue has been marked as a duplicate of 2997 ***</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>

</pre>
</details>

</ul>
</details>

<p>I believe it's the same issue.</p>

<ul>
	<li>
	<ul>
		<li>
		<ul>
			<li>This issue has been marked as a duplicate of 2997 ***</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `duplicates` from Jira: <a href="https://github.com/jenkinsci/jenkins/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-2997%22">JENKINS-2997</a>]</i>
